### ISSUE TITLE
Failed to install js 2.3.4

### ISSUE THREAD

**Issue by manavghaigithub** [getgauge/gauge-js#200]
**Expected**
User should be able to install JS from command line using below command
gauge install js --file gauge-js-2.3.4.zip

**Actual**
User is getting error when trying to install JS 2.3.4

Steps:
1. Download JS 2.3.4
2. Open cmd and run command "gauge install js --file gauge-js-2.3.4.zip"
3. User is getting below error
C:\Users\mghai\Desktop\Gauge>gauge install js --file gauge-js-2.3.4.zip
Failed to execute 'C:\Program Files\nodejs\node.exe C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js configure --fallback-to-build --library=static_library --module=C:\Users\mghai\AppData\Local\Temp\1\gauge_temp1556847522182553100\node_modules\grpc\src\node\extension_binary\node-v64-win32-x64-unknown\grpc_node.node --module_name=grpc_node --module_path=C:\Users\mghai\AppData\Local\Temp\1\gauge_temp1556847522182553100\node_modules\grpc\src\node\extension_binary\node-v64-win32-x64-unknown --napi_version=3 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v64' (1)
Failed to install plugin 'js'.
Reason: exit status 1
Failed to install plugin 'js'.
Reason: exit status 1

**Version**
Gauge version: 1.0.4
Commit Hash: 3a9a647

Plugins
-------
csharp (0.10.5)
dotnet (0.1.2.nightly-2019-02-14)
html-report (4.0.7)
java (0.7.2)
screenshot (0.0.1)

**Comment by nehashri** [comment]
@manavghaigithub What version of node are you using?
I think this issue maybe a duplicate of https://github.com/getgauge/gauge-js/issues/199 if you are using node 12. Please let us know.

**Comment by manavghaigithub** [comment]
@nehashri I am on node 10.4.1
C:\Users\mghai\Desktop\Gauge>node --version
v10.4.1

**Comment by nehashri** [comment]
@manavghaigithub Is there proper internet connection the machine? Is the machine behind some proxy?

**Comment by manavghaigithub** [comment]
@nehashri i am doing it from office so yes there are proxies

**Comment by zabil** [comment]
Yes. Unfortunately it's not truly offline as the the js plugin needs to download a few npm modules.
You'll have to set proxy https://docs.gauge.org/latest/configuration.html#http-proxy and try it.

**Comment by manavghaigithub** [comment]
@zabil The link you shared talks about setting proxy server "export HTTP_PROXY=http://10.0.2.2:5678"

Where do i need to set this environment variable? I am trying to install using below command
"gauge install js --file gauge-js-2.3.4.zip"

**Comment by nehashri** [comment]
@manavghaigithub We have updated the instructions to include windows commands. 
For windows, you need to do `set HTTP_PROXY=http://10.0.2.2:5678`.

**Comment by manavghaigithub** [comment]
Neha - which file do I need to set this env variable?Please provide more details so that I can go back and check with my company’s respective team.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by nehashri** [comment]
@manavghaigithub You have to set this on your machine. Its not on a file, but an environment variable on the machine that you are using. See https://docs.microsoft.com/en-us/windows/desktop/procthread/environment-variables

**Comment by manavghaigithub** [comment]
Oh Okhay got it. Sorry for weird question, I thought you are asking to set it somewhere in the some file. I will try and let you know in case of any further question.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by njbrake** [comment]
Hi everyone, thank you for all the information in this thread! I am also working through a similar problem as @manavghaigithub, except in my case I am trying to set the gauge js plugin up on a completely offline virtual machine: do you have any advice about how to install the js plugin on a system that has no internet access whatsoever? Thanks!

**Comment by manavghaigithub** [comment]
@nehashri @zabil Zabil mentioned its not true offline as it downloads few components from NPM. Can you please provide details on what all modules it downloads from NPM? We need to go back to our concern team on the same.

**Comment by nehashri** [comment]
@manavghaigithub the packages that need to be installed can found here(https://github.com/getgauge/gauge-js/blob/master/package.json#L26), the dependencies mentioned need to be downloaded. The problem is with the `grpc` dependency, its needs to be compiled locally on the machine, hence offline installation is not working right now.

**Comment by nehashri** [comment]
@njbrake No, we haven't explored the completely offline installation of gauge-js because of the `grpc` dependency we have. However, since many of you are facing this issue, I will try to put up some notes on the same soon.

**Comment by sriv** [comment]
For a true offline install, we could use [`npm pack`](https://docs.npmjs.com/cli/pack.html) to create a tarball of gauge-js. 

However this will deviate from the current plugin install mechanism. So this needs some reconciliation. Also, as @nehashri points out, the dependency on `grpc` involves some native binaries, so that would make this platform dependant.

**Comment by nehashri** [comment]
GRPC has support for server with pure javascript now. We need use this implementation of GRPC and pack the dependencies along with gauge-js zip to support offline installation.

**Comment by negiDharmendra** [comment]
I have created a [forked](https://github.com/getgauge/gauge-jso) of Gauge-JS runner which uses the pure js implementation of GRPC. Please give it a try and let us know if you face any issue. 

**Instructions**

- git clone https://github.com/getgauge/gauge-jso.git
- npm install 
- npm run package ( this will create a packaging inside deploy directory)
- gauge install jso -f ./deploy/gauge-jso-2.3.6.zip
- Change the `Language` field of `<path-gauge-project>/manifest.json` to `jso` in your existing GaugeJS project

**Comment by Debashis9012** [comment]
Tested the above fix.
Observation:
while doing `gauge init js` after installing `gauge-js-offline` plugin it shows the below console log:

`Initialising Gauge JavaScript project
{ [Error: ENOENT: no such file or directory, mkdir '/Users/debas/work/structureLogs/js2/test/env/default']
errno: -2,
code: 'ENOENT',
syscall: 'mkdir',
path:
'/Users/debas/work/structureLogs/js2/test/env/default' }
 create  manifest.json
 create  specs
 create  specs/example.spec
 create  .gitignore
 create  env
 create  env/default
 create  env/default/default.properties
Successfully initialized the project. Run specifications with "gauge run specs/".`

Tested version:
```
gauge version: 1.0.7.nightly-2019-09-16
Commit Hash: 047a847

Plugins
-------
js (2.3.6.nightly-2019-09-26)
```

**Comment by negiDharmendra** [comment]
We have created a separate js runner with offline installation support which is based on [grpc-js](https://github.com/grpc/grpc-node/tree/master/packages/grpc-js).

As of now the offline installation support is only available on the [latest nightly version ](https://bintray.com/gauge/gauge-js/Nightly/). Please make sure you download the correct nightly version (note the file pattern `gauge-js-offline-<version number>.nightly-<date>.zip`)

**Note :** The offline JS runner does not work with nodejs version below `10.10.0`

**Comment by negiDharmendra** [comment]
@Debashis9012 Can you please create a separate issue for the [error](https://github.com/getgauge/gauge-js/issues/200#issuecomment-535416688) you were facing. As the error is not specific to the runner with offline installation support.

**Comment by Debashis9012** [comment]
- Successfully installed js plugin in offline mode. 
- @negiDharmendra I have created another [issue](https://github.com/getgauge/gauge-js/issues/252). (when doing gauge init js)
- All the basic features in vs-code for js projects works as expected.

Tested version: 
```
Gauge version: 1.0.7.nightly-2019-09-16
js (2.3.6.nightly-2019-09-24)
```

**Comment by jbadeau** [comment]
Very confusing. Is v2.3.6 supposed to be installable without any internet connection (.e.g. no binary compiling)? What is the purpose of 1.0.7.nightly-2019-09-16? Is that a version using grpc-js?

**Comment by nehashri** [comment]
@jbadeau Sorry for the confusion caused here. 

We have released a new version of [Gauge JS v2.3.6](https://github.com/getgauge/gauge-js/releases/tag/v2.3.6) with the offline installer (gauge-js-offline-2.3.6.zip). The offline zip installer will not have any binary compiling and can be installed without a internet connection.
